apollo/smoothie.py

```diff
@@ -58,70 +24,182 @@ def connect(self):
+            # Handle M112
+            if gcode.startswith('M112') and not is_gcode_done:
+                if response.startswith('ok'):
+                    is_gcode_done = True
+
+            # Handle M119
+            if gcode.startswith('M119') and not is_gcode_done:
+                try:
+                    json_result = json.loads(response)
+                    is_gcode_done = True
+                    logger.info('formatted data'.format(str(json_result)))
+                except ValueError:
+                    pass
+
+            # Handle M199
+            if gcode.startswith('M199') and not is_gcode_done:
+                try:
+                    json_result = json.loads(response)
+                    is_gcode_done = True
+                    logger.info('formatted data'.format(str(json_result)))
+                except ValueError:
+                    pass
+
+            # Handle M204
+            if gcode.startswith('M204') and not is_gcode_done:
+                try:
+                    #status, data = response.split(' ')
+                    json_result = json.loads(response)
+                    is_gcode_done = True
+                    logger.info('formatted data'.format(str(json_result)))
+                except ValueError:
+                    pass
+
+            # Handle: M999
+            if gcode.startswith('M999') and not is_gcode_done:
+                if response == 'ok':
+                    is_gcode_done = True
+
+            if response == '{"stat":0}':
                 break
-            
-            # check whether JSON/dict format
-            if data.find('{') > 0:
-                # pull out any text data at beginning
-                text = data[:data.find('{')]
-                res.append(text)
 
-                # convert to JSON object
-                jtxt = data[data.find('{'):]
-                res.append(jtxt)
-                jobj = json.loads(jtxt)
+        if not (gcode == 'M112'):
+            yield from self.turn_off_feedback()
 
-                # check whether has "stat" and value if so
-                if 'stat' in jobj:
-                    if jobj['stat'] == 0:
-                        done = True
+        return json_result
 
-        if response_handler:
-            response_handler.send(None)
+    @asyncio.coroutine
+    def turn_on_feedback(self):
+        yield from self.send_feedback_gcode('M62\r\n', 'feedback engaged')
 
-        return res
+    @asyncio.coroutine
+    def turn_off_feedback(self):
+        yield from self.send_feedback_gcode('M63\r\n', 'feedback disengaged')
+
+    @asyncio.coroutine
+    def send_feedback_gcode(self, delimited_gcode, expected_response_msg):
+        
+        yield from self.write_and_drain(delimited_gcode)
+
+        # while True needs an escape to avoid infinite loop(s) - how many times does this get called???
+        # counter used to prevent infinite loop, recorded in logger to see if this ever happens
+        # while True:
+        counter = 0
```

Comment:
Maybe rename this variable to `read_attempts` or something a little more descriptive than `counter`